### PR TITLE
Improved defer and go inspections (fixes #2202)

### LINKED ISSUE #2202: Defer recover inspection
In this example:
```go
func demo() {
	defer recover()
}
```
the `defer recover()` is not particularly useful for catching panics. It would be far more useful to have something like this as a quick fix:
```go
func demo() {
	defer func(){
		if err := recover(); err != nil {
			// do something with the panic value
		}
	}()	
}
```

### PR DESCRIPTION
**Pull request by dlsniper**

### CODE REVIEW

**Comment by dlsniper** [comment]
I couldn't get the quickfix test to run on my setup for some reason, I hope it will work in TeamCity. If not, I'd be curious about what's wrong with the way I've done it.

**Comment by dlsniper** [comment]
Also not sure if I should split this up into its own inspection or not.

File: src/com/goide/inspections/GoDeferGoInspection.java [hunk 1]
```diff
@@ -35,17 +39,40 @@ protected GoVisitor buildGoVisitor(@NotNull final ProblemsHolder holder, @NotNul
...
-        checkExpression(o.getExpression(), "go");
+        checkExpression(o, o.getExpression(), "go");
       }
 
-      private void checkExpression(@Nullable GoExpression o, String who) {
-        if (o == null || o instanceof GoCallExpr || o instanceof GoBuiltinCallExpr) return;
+      private void checkExpression(@NotNull GoStatement statement, @Nullable GoExpression o, String who) {
+        if (o == null || o instanceof GoBuiltinCallExpr) return;
+        if (o instanceof GoCallExpr) {
+          if (GoPsiImplUtil.isPanic((GoCallExpr)o)) {
```
Comment (zolotov) [inline src/com/goide/inspections/GoDeferGoInspection.java]: both if-branches looks very similar, any chance to simplify?

File: src/com/goide/psi/impl/GoPsiImplUtil.java [hunk 2]
```diff
@@ -81,8 +81,16 @@ public static boolean builtin(@Nullable PsiElement resolve) {
...
 
   public static boolean isPanic(@NotNull GoCallExpr o) {
+    return stdLibCall(o, "panic");
+  }
+
+  public static boolean isRecover(@NotNull GoCallExpr o) {
+    return stdLibCall(o, "recover");
+  }
+
+  private static boolean stdLibCall(@NotNull GoCallExpr o, String name) {
```
Comment (zolotov) [inline src/com/goide/psi/impl/GoPsiImplUtil.java]: cool

File: src/com/goide/inspections/GoDeferGoInspection.java (hunk 1 above)
Comment (zolotov) [inline src/com/goide/inspections/GoDeferGoInspection.java]: actually, the new control flow is strange a little. Firstly we have two different methods for `go` and `defer` statement and we join them in a single `checkExpression` method. And then we separate them again by conditions like `if ("go".equals(who))`, so why did we join them in the first place?

File: tests/com/goide/quickfix/GoDeferGoQuickFixTest.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    myFixture.enableInspections(GoDeferGoInspection.class);
+  }
+
+  @NotNull
+  @Override
+  protected String getBasePath() {
+    return "quickfixes/defer-go";
+  }
+
+  public void testDeferRecover() { doTest(GoReplaceWithCorrectDeferRecoverQuickFix.QUICK_FIX_NAME, true); }
```
Comment (zolotov) [inline tests/com/goide/quickfix/GoDeferGoQuickFixTest.java]: move to `GoDeferGoInspectionTest`

**Comment by dlsniper** [comment]
I've reworked this but I on local I have the following issue and it seems on TeamCity I can't get the tests to run at all due to some Gradle issues.

I've tried investigating the issue but it looks like ` recover ` doesn't resolve to the SDK `builtin/builtin.go` ` recover ` definition which I can't explain as I can't replicate it on the debug instance.

```

junit.framework.AssertionFailedError
java.lang.RuntimeException: junit.framework.AssertionFailedError
	at com.intellij.openapi.application.impl.LaterInvocator.invokeAndWait(LaterInvocator.java:177)
	at com.intellij.openapi.application.impl.ApplicationImpl.invokeAndWait(ApplicationImpl.java:641)
	at com.intellij.testFramework.EdtTestUtilKt.runInEdtAndWait(EdtTestUtil.kt:42)
	at com.intellij.testFramework.EdtTestUtil$Companion.runInEdtAndWait(EdtTestUtil.kt:29)
	at com.intellij.testFramework.EdtTestUtil.runInEdtAndWait(EdtTestUtil.kt)
	at com.intellij.testFramework.UsefulTestCase.runBare(UsefulTestCase.java:423)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:129)
	at junit.framework.TestSuite.runTest(TestSuite.java:252)
	at junit.framework.TestSuite.run(TestSuite.java:247)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:86)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.runTestClass(JUnitTestClassExecuter.java:114)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.execute(JUnitTestClassExecuter.java:57)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessor.processTestClass(JUnitTestClassProcessor.java:66)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at sun.reflect.GeneratedMethodAccessor72.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:109)
	at sun.reflect.GeneratedMethodAccessor71.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:377)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
	at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: junit.framework.AssertionFailedError
	at junit.framework.Assert.fail(Assert.java:55)
	at junit.framework.Assert.assertTrue(Assert.java:22)
	at junit.framework.Assert.assertNotNull(Assert.java:256)
	at junit.framework.Assert.assertNotNull(Assert.java:248)
	at junit.framework.TestCase.assertNotNull(TestCase.java:417)
	at com.goide.quickfix.GoQuickFixTestBase.applySingleQuickFix(GoQuickFixTestBase.java:41)
	at com.goide.quickfix.GoQuickFixTestBase.doTest(GoQuickFixTestBase.java:34)
	at com.goide.inspections.GoDeferGoInspectionTest.testDeferRecover(GoDeferGoInspectionTest.java:58)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at junit.framework.TestCase.runTest(TestCase.java:176)
	at com.intellij.testFramework.UsefulTestCase.access$001(UsefulTestCase.java:80)
	at com.intellij.testFramework.UsefulTestCase.lambda$runTest$0(UsefulTestCase.java:326)
	at com.intellij.testFramework.EdtTestUtil$Companion$runInEdtAndWait$2.invoke(EdtTestUtil.kt:33)
	at com.intellij.testFramework.EdtTestUtil$Companion$runInEdtAndWait$2.invoke(EdtTestUtil.kt:27)
	at com.intellij.testFramework.EdtTestUtilKt$sam$Runnable$ca4818df.run(EdtTestUtil.kt)
	at com.intellij.openapi.application.impl.ApplicationImpl.invokeAndWait(ApplicationImpl.java:633)
	at com.intellij.testFramework.EdtTestUtilKt.runInEdtAndWait(EdtTestUtil.kt:42)
	at com.intellij.testFramework.EdtTestUtil$Companion.runInEdtAndWait(EdtTestUtil.kt:33)
	at com.intellij.testFramework.EdtTestUtil.runInEdtAndWait(EdtTestUtil.kt)
	at com.intellij.testFramework.UsefulTestCase.invokeTestRunnable(UsefulTestCase.java:357)
	at com.intellij.testFramework.UsefulTestCase.runTest(UsefulTestCase.java:341)
	at com.intellij.testFramework.fixtures.LightPlatformCodeInsightFixtureTestCase.access$001(LightPlatformCodeInsightFixtureTestCase.java:41)
	at com.intellij.testFramework.fixtures.LightPlatformCodeInsightFixtureTestCase.doRunTests(LightPlatformCodeInsightFixtureTestCase.java:129)
	at com.intellij.testFramework.fixtures.LightPlatformCodeInsightFixtureTestCase.runTest(LightPlatformCodeInsightFixtureTestCase.java:121)
	at com.intellij.testFramework.UsefulTestCase.defaultRunBare(UsefulTestCase.java:368)
	at com.intellij.testFramework.EdtTestUtil$Companion$runInEdtAndWait$1.invoke(EdtTestUtil.kt:29)
	at com.intellij.testFramework.EdtTestUtil$Companion$runInEdtAndWait$1.invoke(EdtTestUtil.kt:27)
	at com.intellij.testFramework.EdtTestUtilKt$sam$Runnable$ca4818df.run(EdtTestUtil.kt)
	at com.intellij.openapi.application.TransactionGuardImpl$6.run(TransactionGuardImpl.java:291)
	at com.intellij.openapi.application.impl.LaterInvocator$1.run(LaterInvocator.java:156)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.runNextEvent(LaterInvocator.java:326)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.run(LaterInvocator.java:310)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:756)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:76)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:726)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:863)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:699)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:385)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)

```

**Comment by zolotov** [comment]
there is a very special SDK mock in tests, see `mockSdk-1.1.2` and feel free to add anything you miss for tests

**Comment by zolotov** [comment]
It's even simpler, looks like you haven't attached sdk to test. See any `*SdkAware` test as an example

**Comment by dlsniper** [comment]
done, thank you!

File: src/com/goide/inspections/GoDeferGoInspection.java [hunk 4]
```diff
@@ -36,20 +38,58 @@ protected GoVisitor buildGoVisitor(@NotNull final ProblemsHolder holder, @NotNul
...
+          }
+        }
+
         checkExpression(o.getExpression(), "go");
       }
 
       private void checkExpression(@Nullable GoExpression o, String who) {
-        if (o == null) return;
-        if ((o instanceof GoCallExpr || o instanceof GoBuiltinCallExpr) && !GoPsiImplUtil.isConversionExpression(o)) {
+        if (o == null || o instanceof GoBuiltinCallExpr) return;
```
Comment (zolotov) [inline src/com/goide/inspections/GoDeferGoInspection.java]: why?

**Comment by zolotov** [comment]
merged manually

File: src/com/goide/inspections/GoDeferGoInspection.java (hunk 4 above)
Comment (dlsniper) [inline src/com/goide/inspections/GoDeferGoInspection.java]: I would like to to have a better solution for that case, and usually people don't `defer new()`. Probably I should change it do not ignore it if the function call is for `recover()` in `defer` calls.

Comment (zolotov) [inline src/com/goide/inspections/GoDeferGoInspection.java]: I mean why you excluded GoBuiltinCallExpr?

**Pull request closed without merging**